openmdao/docs/openmdao_book/features/building_blocks/math_library/math_library.ipynb

```diff
@@ -0,0 +1,601 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ac28484d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from functools import partial\n",
+    "\n",
+    "import numpy as np\n",
+    "import jax\n",
+    "import jax.numpy as jnp\n",
+    "\n",
+    "import openmdao.api as om\n",
+    "from openmdao.math import act_tanh\n",
+    "\n",
+    "\n",
+    "@om.register_jax_component\n",
+    "class RootMeanSquareSwitchComp(om.ExplicitComponent):\n",
+    "    \n",
+    "    def initialize(self):\n",
+    "        self.options.declare('vec_size', types=(int,))\n",
+    "        self.options.declare('mu', types=(float,), default=0.01)\n",
+    "        self.options.declare('threshold', types=(float,), default=0.5)\n",
+    "        \n",
+    "        # This option is only used for this demonstration.\n",
+    "        # The user only needs to implement the partials calculation method\n",
+    "        # that makes sense in their applicaiton.\n",
+    "        self.options.declare('partials_method',\n",
+    "                             values=('jacfwd', 'jacrev', 'jvp', 'vjp'))\n",
+    "    \n",
+    "    def setup(self):\n",
+    "        n = self.options['vec_size']\n",
+    "        self.add_input('x', shape=(n,))\n",
+    "        self.add_output('rms', shape=(1,))\n",
+    "        self.add_output('rms_switch', shape=(1,))\n",
+    "        \n",
+    "        # The partials are a dense row in this case (1 row x N inputs)\n",
+    "        # There is no need to specify a sparsity pattern.\n",
+    "        self.declare_partials(of=['rms', 'rms_switch'], wrt=['x'])\n",
+    "\n",
+    "        self._partials_method = {'jacfwd': self._compute_partials_jacfwd,\n",
+    "                                 'jacrev': self._compute_partials_jacrev,\n",
+    "                                 'jvp': self._compute_partials_jvp,\n",
+    "                                 'vjp': self._compute_partials_vjp}\n",
+    "\n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jacfwd(self, x):\n",
+    "        deriv_func = jax.jacfwd(self.compute_primal, argnums=[0])\n",
+    "        # Always returns a tuple\n",
+    "        drms_dx, dswitch_dx = deriv_func(x)\n",
+    "        return drms_dx, dswitch_dx\n",
+    "    \n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jacrev(self, x):\n",
+    "        deriv_func = jax.jacrev(self.compute_primal, argnums=[0])\n",
+    "        # Always returns a tuple\n",
+    "        drms_dx, dswitch_dx = deriv_func(x)\n",
+    "        return drms_dx, dswitch_dx\n",
+    "\n",
+    "    @partial(jax.jit, static_argnums=(0,))\n",
+    "    def _compute_partials_jvp(self, x):\n",
+    "        # Note that JVP is a poor choice here, since the jacobian is a row vector!\n",
+    "        \n",
+    "        # Multipling the jacobian by a column vector of ones results in a scalar output,\n",
```

Comment:
While a user *could* compute a full jacobian using jvp or vjp if they wanted to, I'm not sure that we should be showing them an example of how to do it that way vs. how to use jvp/vjp with `compute_jacvec_product`.

Reply:
I understand that you retracted this below but I thought it was valuable to show them how it works conceptually.  I'll include another example with a diagonal jacobian and a single column jacobian to try to explain in more depth.

Reply:
The part I was retracting below was the 'Doesn't the jacobian have 2 rows' comment, not the part about using jvp or vjp to compute the whole jacobian.  I don't think we should be showing users how to do something that we can get away with in this special case because the jacobian is small enough where it doesn't matter, but in general could be very inefficient.  Showing an example there jvp/vjp are called from compute_jacvec_product would make more sense I think.